pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,185 @@
+                try {
+                    while (isRunning) {
+                        final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);
+                        DataPacket dataPacket = transaction.receive();
+                        if (dataPacket == null) {
+                            // no data available. Wait a bit and try again
+                            try {
+                                Thread.sleep(waitTimeMs);
+                            } catch (InterruptedException e) {
+                                log.warn("Failed to thread sleep milliseconds " + waitTimeMs, e);
+                            }
+                            continue;
+                        }
+
+                        final List<NiFiDataPacket> dataPackets = Lists.newArrayList();
+                        do {
+                            // Read the data into a byte array and wrap it along with the attributes
+                            // into a NiFiDataPacket.
+                            final InputStream inStream = dataPacket.getData();
+                            final byte[] data = new byte[(int) dataPacket.getSize()];
+                            StreamUtils.fillBuffer(inStream, data);
+
+                            final Map<String, String> attributes = dataPacket.getAttributes();
+                            final NiFiDataPacket NiFiDataPacket = new StandardNiFiDataPacket(data, attributes);
+                            dataPackets.add(NiFiDataPacket);
+                            dataPacket = transaction.receive();
+                        } while (dataPacket != null);
+
+                        // Confirm transaction to verify the data
+                        transaction.confirm();
+
+                        for (NiFiDataPacket dp : dataPackets) {
+                            consume(new NiFiRecord(dp, transaction));
+                        }
+
+                        transaction.complete();
+                    }
+                } finally {
+                    try {
+                        client.close();
+                    } catch (final IOException ioe) {
+                        log.warn("Failed to close client", ioe);
+                    }
+                }
+            } catch (final IOException ioe) {
+                log.warn("Failed to receive data from NiFi", ioe);
+            }
+        }
+    }
+
+    static private class NiFiRecord implements Record<NiFiDataPacket>{
+        private final NiFiDataPacket value;
+        private final Transaction transaction;
+
+        public NiFiRecord(NiFiDataPacket value, Transaction transaction) {
+            this.value = value;
+            this.transaction = transaction;
+        }
+
+        @Override
+        public NiFiDataPacket getValue() {
+            return value;
+        }
+
+        @Override
+        public void ack() {
+            try {
+                transaction.complete();
```

Comment:
a transaction has multiple records from NiFi. you shouldn't complete when any record acks. You can only complete when all the records of this transaction ack.

Reply:
this still isn't correct. because a transaction has multiple records, you can complete the transaction when a record is acked. you need to track and make sure all the records have been acked before completing the transaction.